File: code/C++/libraries/ADCDevice/ADCDevice.cc
//
// Filename:    ADCDevice.cc
// Description: Cleaned up version of Freenove ADC Module library (C++)
//
////////////////////////////////////////////////////////////////////////////////////////

////////////////////////////////////////////////////////////////////////////////////////
//
// Pre-processor directives
//
#include "ADCDevice.h"

////////////////////////////////////////////////////////////////////////////////////////
//
// ADCDevice constructor
//
ADCDevice::ADCDevice(){
    address = 0;
    wiringPiSetup();
}

////////////////////////////////////////////////////////////////////////////////////////
//
// Attempt to detect which ADC module is in use
//
int ADCDevice::detectI2C(int  addr)
{
    ////////////////////////////////////////////////////////////////////////////////////
    //
    // Initialize wiringPi I2C subsystem
    //
    _fd                              = wiringPiI2CSetup (addr);

    ////////////////////////////////////////////////////////////////////////////////////
    //
    // Check for failure of initializing wiringPi I2C subsystem
    //
    if (_fd                         <  0)
    {
        printf("Error address : 0x%x \n",addr);
        return 0;
    }
    ////////////////////////////////////////////////////////////////////////////////////
    //
    // Attempt writing to the device, checking for a failure
    //
    else if(wiringPiI2CWrite(_fd,0) <  0)
    {
        printf("Not found device in address 0x%x \n",addr);
        return 0;

    }
    ////////////////////////////////////////////////////////////////////////////////////
    //
    // If writing to the device was successful, we have our module
    //
    else
    {
        printf("Found device in address 0x%x \n",addr);
        return 1;
    }
}

////////////////////////////////////////////////////////////////////////////////////////
//
// PCF8591 constructor
//
PCF8591::PCF8591(int addr)
{
    ////////////////////////////////////////////////////////////////////////////
    //
    // Initialize ADCDevice attributes to PCF8591 settings
    //
    address = addr;
    cmd     = 0x40; //The default command for PCF8591 is 0x40.


    ////////////////////////////////////////////////////////////////////////////
    //
    // Establish analogRead and analogWrite functions per detected ADC module
    //
    wiringPiSetup();
    detectI2C(address);

    printf("PCF8591 setup successful! \n");
}

////////////////////////////////////////////////////////////////////////////////////////
//
// a PCF8591-specific analogRead function
//
int  PCF8591::analogRead (int  chn)
{
    wiringPiI2CWrite (_fd, cmd + chn);
    wiringPiI2CRead  (_fd);
    wiringPiI2CWrite (_fd, cmd + chn);
    return (wiringPiI2CRead (_fd));
}

////////////////////////////////////////////////////////////////////////////////////////
//
// a PCF8591-specific analogWrite function
//
int  PCF8591::analogWrite (int  value)
{
    return (wiringPiI2CWriteReg8 (_fd, cmd, value));
}

////////////////////////////////////////////////////////////////////////////////////////
//
// ADS7830 constructor
//
ADS7830::ADS7830(int addr)
{
    ////////////////////////////////////////////////////////////////////////////
    //
    // Initialize ADCDevice attributes to ADS7830 settings
    //
    address = addr;
    cmd     = 0x84; //The default command for is ADS7830 0x40.


    ////////////////////////////////////////////////////////////////////////////
    //
    // Establish analogRead and analogWrite functions per detected ADC module
    //
    wiringPiSetup();
    detectI2C(address);

    printf("ADS7830 setup successful! \n");
}

////////////////////////////////////////////////////////////////////////////////////////
//
// a ADS7830-specific analogRead function
//
int  ADS7830::analogRead (int  chn)
{
    wiringPiI2CWrite (_fd, cmd + chn);
    wiringPiI2CRead  (_fd);
    wiringPiI2CWrite (_fd, cmd + chn);
    return (wiringPiI2CRead (_fd));
}

File: code/python/07.1.1_ADC/ADCDevice.py
#!/usr/bin/env python3
#############################################################################################
# Filename    : ADCDevice.py
# Description : Freenove ADC Module library.
# Author      : www.freenove.com
# modification: 2023/05/11
#############################################################################################

#############################################################################################
##
## import the smbus library
##
import smbus

#############################################################################################
##
## for storing ADCDevice values
##
class ADCDevice(object):

    #########################################################################################
    ##
    ## initializes the attributes of the class
    ##
    def __init__(self):
        self.cmd = 0
        self.address = 0
        self.bus=smbus.SMBus(1)
        # print("ADCDevice init")
        
    #########################################################################################
    ##
    ## detects the device in the provided address
    ##
    def detectI2C(self,addr):
        
        #####################################################################################
        ##
        ## detects the device and displays a conformation message
        ##
        try:
            self.bus.write_byte(addr,0)
            print("Found device in address 0x%x"%(addr))
            return True
            
        #####################################################################################
        ##
        ## displays this message if no device is found
        ##
        except:
            print("Not found device in address 0x%x"%(addr))
            return False
           
    def close(self):
        
        #####################################################################################
        ##
        ## close the connection to the bus pin
        ## 
        self.bus.close()

#############################################################################################
##
## for storing the values for the ADS7830 ADC Module
##        
class PCF8591(ADCDevice):

    #########################################################################################
    ##
    ## initializes the attributes of the class
    ##
    def __init__(self):
        super(PCF8591, self).__init__()
        self.cmd = 0x40     # The default command for PCF8591 is 0x40.
        self.address = 0x48 # 0x48 is the default i2c address for PCF8591 Module.
     
    #########################################################################################
    ##
    ## returns the ADC value from the specified address, command, and pins
    ##   
    def analogRead(self, chn): # PCF8591 has 4 ADC input pins, chn:0,1,2,3
        value = self.bus.read_byte_data(self.address, self.cmd+chn)
        value = self.bus.read_byte_data(self.address, self.cmd+chn)
        return value
    
#############################################################################################
##
## for storing the values for the ADS7830 ADC Module
##
class ADS7830(ADCDevice):

    #########################################################################################
    ##
    ## initializes the attributes of the class
    ##
    def __init__(self):
        super(ADS7830, self).__init__()
        self.cmd = 0x84
        self.address = 0x4b # 0x4b is the default i2c address for ADS7830 Module.   
    
    #########################################################################################
    ##
    ## returns the ADC value from the specified address, command, and pins
    ##    
    def analogRead(self, chn): # ADS7830 has 8 ADC input pins, chn:0,1,2,3,4,5,6,7
        value = self.bus.read_byte_data(self.address, self.cmd|(((chn<<2 | chn>>1)&0x07)<<4))
        return value
